arch/common/shared_irq.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/sw_isr_table.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/spinlock.h>
+
+struct shared_irq_client {
+	void (*routine)(const void *arg);
+	const void *arg;
+};
+
+struct shared_irq_data {
+	bool enabled;
+	struct shared_irq_client clients[CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS];
+	uint32_t client_num;
+	struct k_spinlock lock;
+};
+
+static struct shared_irq_data shared_irq_table[CONFIG_NUM_IRQS];
+
+void shared_irq_isr(const void *data)
+{
+	int i;
+	const struct shared_irq_data *irq_data = data;
+
+	for (i = 0; i < irq_data->client_num; i++)
+		/* note: due to how the IRQ disconnect function works
+		 * this should be impossible but it's better to safe than
+		 * sorry I guess.
+		 */
+		if (irq_data->clients[i].routine)
+			irq_data->clients[i].routine(irq_data->clients[i].arg);
+}
+
+void z_isr_install(unsigned int irq, void (*routine)(const void *),
+		const void *param)
+{
+	struct shared_irq_data *irq_data;
+	unsigned int table_idx;
+	int i;
+	k_spinlock_key_t key;
+
+	table_idx = irq - CONFIG_GEN_IRQ_START_VECTOR;
+
+	/* check for out of bounds table index */
+	if (table_idx >= CONFIG_NUM_IRQS)
+		return;
+
+	irq_data = &shared_irq_table[table_idx];
+
+	/* seems like we've reached the limit of clients */
+	/* TODO: is there a better way of letting the user know this is the case? */
+	if (irq_data->client_num == CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS)
+		return;
+
+	key = k_spin_lock(&irq_data->lock);
+
+	if (!irq_data->enabled) {
```

Comment:
Just test `if(_sw_isr_table[table_idx].isr == shared_irq_isr)` instead and save the space in the irq_data struct.  Actually IMHO having it present and calling it "enabled" is a net loss, because I read that as "interrupt line is enabled" and not "shared interrupts are enabled on this line" and was really confused about what was happening here.